dart/dynamics/BodyNode.cpp

```diff
@@ -888,34 +960,16 @@ void BodyNode::init(Skeleton* _skeleton)
 }
 
 //==============================================================================
-void BodyNode::draw(renderer::RenderInterface* _ri,
-                    const Eigen::Vector4d& _color,
-                    bool _useDefaultColor,
-                    int _depth) const
+void BodyNode::processNewEntity(Entity* _newChildEntity)
 {
-  if (_ri == NULL)
+  if(find(mChildBodyNodes.begin(), mChildBodyNodes.end(), _newChildEntity) !=
+     mChildBodyNodes.end())
```

Comment:
Actually, BodyNode::processNewEntity(~) gets called by Entity::changeParentFrame(~). This means that it will get called for *any* Entity that gets added as a child of the BodyNode.

For example, suppose I make SimpleFrame instance, and I want its parent Frame to be a BodyNode. We'll call Entity::changeParentFrame(~) on the SimpleFrame instance. This will add the SimpleFrame to the mChildEntities vector of the BodyNode. However, the mChildEntities vector contains child BodyNodes as well as all other types of Entities that are children of the BodyNode. Entity::changeParentFrame(~) does not know that the BodyNodes want to sort the BodyNodes out from the rest of the Entities, so we have it call processNewEntity(~). If we were adding a child BodyNode, then it would already exist in mChildBodyNodes, because it would have been added to that vector when we called BodyNode::addChildBodyNode(~). So we check if it's already in there. If it *is* in there, then we quit, because it's already sorted out. If it's *not* in there, then we check if it's already been added to the mNonBodyNodeEntities vector, which is where we want it to be. If it's not already added, then we add it to that vector.

This would probably be an easier thing to explain if I could draw a diagram.

However, I'm glad that you brought my attention to these vectors, because I just realized that we do not yet correctly process child Entities that are being removed from BodyNodes, so I'm going to go take care of that now.